File: packages/core-java/src/main/java/io/github/mertc0604/converttools/coordinates/grid/Gars.java
package io.github.mertc0604.converttools.coordinates.grid;

import io.github.mertc0604.converttools.coordinates.model.AngularCell;
import io.github.mertc0604.converttools.coordinates.model.CoordinatePoint;

import java.util.Locale;
import java.util.regex.Pattern;

public final class Gars {
    private static final String ALPHABET = "ABCDEFGHJKLMNPQRSTUVWXYZ";
    private static final Pattern VALUE = Pattern.compile(
            "^\\d{3}[A-HJ-NP-Z]{2}(?:[1-4](?:[1-9])?)?$"
    );

    private Gars() {
    }

    public static String encode(CoordinatePoint point) {
        return encode(point.longitude(), point.latitude());
    }

    public static String encode(double longitude, double latitude) {
        new CoordinatePoint(latitude, longitude);
        double longitudePosition = (longitude + 180) / 0.5;
        double latitudePosition = (latitude + 90) / 0.5;
        int longitudeIndex = Math.min(
                719,
                (int) Math.floor(longitudePosition)
        );
        int latitudeIndex = Math.min(
                359,
                (int) Math.floor(latitudePosition)
        );
        String longitudeBand = String.format(
                Locale.ROOT,
                "%03d",
                longitudeIndex + 1
        );
        String latitudeBand = new String(new char[]{
                ALPHABET.charAt(latitudeIndex / 24),
                ALPHABET.charAt(latitudeIndex % 24)
        });

        double longitudeRemainder = Math.min(
                1,
                Math.max(0, longitudePosition - longitudeIndex)
        );
        double latitudeRemainder = Math.min(
                1,
                Math.max(0, latitudePosition - latitudeIndex)
        );
        int quadrantColumn = Math.min(
                1,
                (int) Math.floor(longitudeRemainder * 2)
        );
        int quadrantRow = Math.min(
                1,
                (int) Math.floor(latitudeRemainder * 2)
        );
        int quadrant;
        if (quadrantRow == 1) {
            quadrant = quadrantColumn == 0 ? 1 : 2;
        } else {
            quadrant = quadrantColumn == 0 ? 3 : 4;
        }
        double longitudeInQuadrant =
                longitudeRemainder * 2 - quadrantColumn;
        double latitudeInQuadrant =
                latitudeRemainder * 2 - quadrantRow;
        int keypadColumn = Math.min(
                2,
                (int) Math.floor(longitudeInQuadrant * 3)
        );
        int keypadRowFromSouth = Math.min(
                2,
                (int) Math.floor(latitudeInQuadrant * 3)
        );
        int keypad = (2 - keypadRowFromSouth) * 3
                + keypadColumn
                + 1;
        return longitudeBand + latitudeBand + quadrant + keypad;
    }

    public static AngularCell decode(String value) {
        String source = value.toUpperCase(Locale.ROOT)
                .replaceAll("\\s+", "");
        if (!VALUE.matcher(source).matches()) {
            throw new IllegalArgumentException(
                    "Invalid GARS coordinate."
            );
        }

        int longitudeBand = Integer.parseInt(source.substring(0, 3));
        int latitudeBand = ALPHABET.indexOf(source.charAt(3)) * 24
                + ALPHABET.indexOf(source.charAt(4));
        if (longitudeBand < 1
                || longitudeBand > 720
                || latitudeBand < 0
                || latitudeBand > 359) {
            throw new IllegalArgumentException("Invalid GARS band.");
        }

        double longitude = -180 + (longitudeBand - 1) * 0.5;
        double latitude = -90 + latitudeBand * 0.5;
        double cellDegrees = 0.5;
        if (source.length() >= 6) {
            int quadrant = Character.digit(source.charAt(5), 10);
            longitude += (quadrant == 2 || quadrant == 4 ? 1 : 0)
                    * 0.25;
            latitude += (quadrant == 1 || quadrant == 2 ? 1 : 0)
                    * 0.25;
            cellDegrees = 0.25;
        }
        if (source.length() == 7) {
            int keypadIndex = Character.digit(source.charAt(6), 10) - 1;
            longitude += keypadIndex % 3 / 12.0;
            latitude += (2 - keypadIndex / 3) / 12.0;
            cellDegrees = 1.0 / 12;
        }

        return new AngularCell(
                new CoordinatePoint(
                        latitude + cellDegrees / 2,
                        longitude + cellDegrees / 2
                ),
                source.length(),
                cellDegrees
        );
    }
}

File: packages/core-js/src/geodesy/grids/gars.js
import { pointFromValues, validatePoint } from "../core/numbers.js";

const ALPHABET = "ABCDEFGHJKLMNPQRSTUVWXYZ";

export function encodeGars(longitudeValue, latitudeValue) {
  const { latitude, longitude } = pointFromValues(latitudeValue, longitudeValue);
  const safeLongitude = longitude === 180 ? 180 - Number.EPSILON * 128 : longitude;
  const safeLatitude = latitude === 90 ? 90 - Number.EPSILON * 64 : latitude;
  const longitudePosition = (safeLongitude + 180) / 0.5;
  const latitudePosition = (safeLatitude + 90) / 0.5;
  const longitudeIndex = Math.floor(longitudePosition);
  const latitudeIndex = Math.floor(latitudePosition);
  const longitudeBand = String(longitudeIndex + 1).padStart(3, "0");
  const latitudeBand =
    ALPHABET[Math.floor(latitudeIndex / 24)] + ALPHABET[latitudeIndex % 24];

  const longitudeRemainder = longitudePosition - longitudeIndex;
  const latitudeRemainder = latitudePosition - latitudeIndex;
  const quadrantColumn = Math.min(1, Math.floor(longitudeRemainder * 2));
  const quadrantRow = Math.min(1, Math.floor(latitudeRemainder * 2));
  const quadrant =
    quadrantRow === 1
      ? quadrantColumn === 0
        ? 1
        : 2
      : quadrantColumn === 0
        ? 3
        : 4;
  const longitudeInQuadrant = longitudeRemainder * 2 - quadrantColumn;
  const latitudeInQuadrant = latitudeRemainder * 2 - quadrantRow;
  const keypadColumn = Math.min(2, Math.floor(longitudeInQuadrant * 3));
  const keypadRowFromSouth = Math.min(2, Math.floor(latitudeInQuadrant * 3));
  const keypad = (2 - keypadRowFromSouth) * 3 + keypadColumn + 1;
  return `${longitudeBand}${latitudeBand}${quadrant}${keypad}`;
}

export function decodeGars(value) {
  const source = String(value).toUpperCase().replace(/\s+/g, "");
  if (!/^\d{3}[A-HJ-NP-Z]{2}(?:[1-4](?:[1-9])?)?$/.test(source)) {
    throw new Error("Invalid GARS coordinate.");
  }

  const longitudeBand = Number(source.slice(0, 3));
  const latitudeBand =
    ALPHABET.indexOf(source[3]) * 24 + ALPHABET.indexOf(source[4]);
  if (
    longitudeBand < 1 ||
    longitudeBand > 720 ||
    latitudeBand < 0 ||
    latitudeBand > 359
  ) {
    throw new Error("Invalid GARS band.");
  }

  let longitude = -180 + (longitudeBand - 1) * 0.5;
  let latitude = -90 + latitudeBand * 0.5;
  let cellDegrees = 0.5;
  if (source.length >= 6) {
    const quadrant = Number(source[5]);
    longitude += (quadrant === 2 || quadrant === 4 ? 1 : 0) * 0.25;
    latitude += (quadrant === 1 || quadrant === 2 ? 1 : 0) * 0.25;
    cellDegrees = 0.25;
  }
  if (source.length === 7) {
    const keypadIndex = Number(source[6]) - 1;
    longitude += (keypadIndex % 3) / 12;
    latitude += (2 - Math.floor(keypadIndex / 3)) / 12;
    cellDegrees = 1 / 12;
  }

  return {
    ...validatePoint(
      latitude + cellDegrees / 2,
      longitude + cellDegrees / 2,
    ),
    precision: source.length,
    cellDegrees,
  };
}
